rmw_fastrtps_shared_cpp/src/rmw_wait.cpp

```diff
@@ -27,8 +27,85 @@
 #include "fastdds/dds/core/condition/GuardCondition.hpp"
 #include "fastdds/dds/subscriber/DataReader.hpp"
 
+#include "tracy/Tracy.hpp"
+
 namespace rmw_fastrtps_shared_cpp
 {
+
+bool should_skip_wait(
+  rmw_subscriptions_t * subscriptions,
+  rmw_guard_conditions_t * guard_conditions,
+  rmw_services_t * services,
+  rmw_clients_t * clients,
+  rmw_events_t * events)
+{
+  ZoneScoped;
+
+  // Events and guard conditions are the cheapest to check, so check them first
+  if (guard_conditions) {
+    for (size_t i = 0; i < guard_conditions->guard_condition_count; ++i) {
+      void * data = guard_conditions->guard_conditions[i];
+      auto guard_condition = static_cast<eprosima::fastdds::dds::GuardCondition *>(data);
+
+      if (guard_condition->get_trigger_value())
+        return true;
+    }
+  }
+
+  if (events) {
+    for (size_t i = 0; i < events->event_count; ++i) {
+      auto event = static_cast<rmw_event_t *>(events->events[i]);
+      auto custom_event_info = static_cast<CustomEventInfo *>(event->data);
+      if (custom_event_info->get_listener()->get_statuscondition().get_trigger_value())
```

Comment:
Note that events also have a guard condition:

```suggestion
      if (custom_event_info->get_listener()->get_statuscondition().get_trigger_value() ||
          custom_event_info->get_listener()->get_event_guard(event->event_type).get_trigger_value())
```